src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+
+    def _set_treedict(self, in_dict, key, val):
+        """
+        _set_treedict(d, "a.b.c", 42) is like d[a][b][c] = 42
+        but levels are created on demand.
+        """
+        path = key.split(".", 1)
+        if len(path) == 2:
+            self._set_treedict(in_dict.setdefault(path[0], {}), path[1], val)
+        else:
+            in_dict[key] = val
+
+    def setattr(self, key, val):
+        """
+        set job attribute
+        """
+        self._set_treedict(self.jobspec, "attributes." + key, val)
+
+    def setattr_shopt(self, key, val):
+        """
+        set job attribute: shell option
+        """
+        self.setattr("system.shell.options." + key, val)
+
+    def dumps(self):
+        return json.dumps(self.jobspec)
+
+    @property
+    def resources(self):
+        return self.jobspec.get("resources", None)
+
+    @property
+    def tasks(self):
+        return self.jobspec.get("tasks", None)
+
+    @property
+    def attributes(self):
+        return self.jobspec.get("attributes", None)
+
+    @property
+    def version(self):
+        return self.jobspec.get("version", None)
+
+    def resource_walk(self):
+        """
+        Traverse the resources in the `resources` section of the jobspec.
+
+        Performs a depth-first, pre-order traversal.
+        """
+
+        def walk_helper(res_list):
+            for resource in res_list:
+                yield resource
+                children = resource.get("with", [])
+                # TODO: convert to `yield from` after dropping 2.7
+                for res in walk_helper(children):
+                    yield res
+
+        return walk_helper(self.resources)
+
+    def resource_walk_type(self, res_type):
+        """
+        Traverse the resources of type `res_type` in the `resources` section of the
+        jobspec.
+
+        Performs a depth-first, pre-order traversal.
+
+        :param res_type: resource type(s) to filter by
+        :type res_type: string or iterable of strings
+        """
+        if isinstance(res_type, six.string_types):
+            func = lambda x: x["type"] == res_type
+        elif isinstance(res_type, collectionsAbc.Iterable):
+            func = lambda x: x["type"] in res_type
+        else:
+            raise TypeError("res_type must be a string or an iterable")
+        return (x for x in self.resource_walk() if func(x))
+
+    def count_num_type(self, res_type):
+        def _count_helper(curr_res, res_type):
+            resource_count = curr_res["count"]
+            if curr_res["type"] == res_type:
+                return resource_count
+            num_in_children = sum(
+                _count_helper(r, res_type) for r in curr_res.get("with", [])
+            )
+            return resource_count * num_in_children
+
+        return sum(_count_helper(r, res_type) for r in self.jobspec["resources"])
+
+    def submit(self, flux_handle, flags=0):
+        return submit(flux_handle, self.dumps(), flags)
```

Comment:
This seems a bit odd to me, which submit does this end up invoking? Why this here rather than as a method on the handle or a function elsewhere? I get why it's kinda convenient, but feels out of place.

Reply:
Maybe `job.submit()` (and `submit_async()`) could take either the jobspec string as it does now or a `Jobspec()` object?     Then drop this method?